proseco/guide.py

```diff
@@ -153,13 +161,49 @@ def run_search_stages(self):
                 self.log(f'Quitting after stage {idx - 1} with {already_selected} stars', level=1)
                 break
         self.log('Done with search stages')
-        selected = cand_guides[cand_guides['stage'] != -1]
-        selected.sort(['stage', 'mag'])
-        if len(selected) >= self.n_guide:
-            return selected[0:n_guide]
+        stage_cands = cand_guides[cand_guides['stage'] != -1]
+        stage_cands.sort(['stage', 'mag'])
+        selected = self.select_catalog(stage_cands[0:n_guide + GUIDE.surplus_stars])
         if len(selected) < self.n_guide:
             self.log(f'Could not find {self.n_guide} candidates after all search stages')
-            return selected
+        return selected
+
+    def select_catalog(self, stage_cands):
+        """
+        For the candidates selected at any stage (stage-assigned candidates?) select the
+        first combination that satisfies the most number of extra tests.
+        Here the extra tests are just the 3 checks for guide stars that are
+        too tightly clustered.
+        """
+        self.log(f'Selecting catalog from {len(stage_cands)} stage-selected stars')
+        # Set some arrays to save "check status" and combination identifiers so
+        # we only have to search through the candidate combinations ones.  The plan is
+        # if no combination satisfies all the checks, lower the standard (required success
+        # on n checks) and repeat, but do this just searching through the previously
+        # determined status.
+        saved_status = []
+        saved_idxs = []
+
+        # I should come back to this and see if the "min" is needed or if the combinations
+        # code runs fine even if we have fewer-than-expected stage_cands to start
+        choose_n = min(len(stage_cands), self.n_guide)
+
+        for comb in combinations(range(len(stage_cands)), choose_n):
+            cluster_status = run_comb_checks(stage_cands[[comb]])
```

Comment:
This code is advertised above to be extensible for other checks, and is named in a general way, but the surrounding code here is hardwired to just the cluster checks.  (Everything will break if you add a 4th test).  I think it should be more like:
```
select_results = {}
for comb in index_combinations(len(stage_cands), choose_m):   # Usually use "m", as in N choose M.
    cands = stage_cands[list(comb)]  # (note that [(1,2)] is not the same as list((1,2))
    n_pass, n_tests = run_select_checks(cands)  # This function knows how many tests get run
    if n_pass == n_tests:
        return cands
    else:
        if n_pass not in select_results:
            # Keep a copy of the first cands table with n_pass passing tests
            select_results[n_pass] = cands

if len(select_results) == 0:
    # Fall back to the original
else:
    # Return the catalog with the maximum n_pass key value
    return select_results[max(select_results)]
```

Reply:
If you can't satisfy all 3 tests, you still want the first combination (in your preferred order) that would satisfy 2 of 3.  How is the "first combination in your preferred order" part guaranteed with the use of the select_results dict?

Reply:
Oh wait, I get it, you win that as part of "if n_pass not in select_results".

Reply:
Thought I'm still not sure how we'd ever get to "if len(select_results) == 0"? n_pass == 0 is a valid key.

Reply:
Oops that's right, you always get an n_pass.  So it's even simpler!  Just `return select_results[max(select_results)]` at the end without the conditional block.

Reply:
:+1:   

Though I think for belt-and-suspenders on weird cases, the modified run_select_checks should be ready for the case where no tests could be run (I think this is just the only 3-or-fewer-candidates-in-the-field case) and should just return 0, 0 in that case instead of unhandled exception.

Reply:
Well `n_tests` should still be whatever it is.  There is no a priori requirement that if one case "cannot run tests", that this will apply to subsequent combinations.  The only thing you lose by letting it continue is a tiny performance hit.  And in fact you might want to return -1 for n_pass (no tests were able to run), which is worse than 0 (no tests passed).

Reply:
Yep, I just wasn't sure if I had to handle it differently, but the code that is there (with the list of statuses) will basically work.  I think I'll leave the -1 as an idea for RFE.